### ISSUE TITLE
Support for Interval (Duration) data types

### ISSUE THREAD

**Issue by mgabeler-lee-6rs** [facebook/ent#1168]
Support for PostgreSQL (and others' equivalent) `interval` data type, mapping to a Go `time.Duration`, would be quite handy.

- [x] I have searched the [issues](https://github.com/facebook/ent/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Would like to be able to write something like the following in my schema `.go` file:
```go
func (MyEntity) Fields() []ent.Field {
	return []ent.Field{
		// ...
		field.Duration("myIntervalProperty"),
		// ...
	}
}
```

For DB dialects that have a data type for this, it would use that (e.g. PG's `interval` type as above). For other dialects, it could use an idiomatic mapping for that DB, or fall back on a (consistent) string or numeric representation.

## Motivation 🔦

Similar motivations to mapping time(stamp)s as a typed entity, instead of as a string or a number of (milli/micro/nano)seconds since an epoch:
* Avoids confusion around units -- if storing an interval as a raw number, it's easy to accidentally interpret it as the wrong units
* Allows using database-native operations on the value -- e.g. can ask the database to add an interval to a timestamp, initialize an interval as the difference between two timestamps, and use either of those in (in)equality queries
  * Concrete use case: I have an entity that needs a "TTL for children" attribute, with semantics along the lines of "when creating a new child, set its expiration time(stamp) to `now + parent.childrenTTL`"
* Allows accurate reflection of existing database schemas
  * If I'm trying to add an ent mapping for a pre-existing database, it's unclear how I could map it in ent. I suspect I could tell ent it's a string, but Go doesn't seem to have any built-in code for parsing durations in custom formats, so this gets yucky fast

**Comment by a8m** [comment]
Hey @mgabeler-lee-6rs and thanks for proposing this option.

It's possible to define [`GoType`](https://entgo.io/docs/schema-fields/#go-type) for schema fields and as an example you can map `time.Duration` to its actual type (`int64`) as follows:

```go
field.Int64("duration").
	GoType(time.Duration(0)).
	Optional(),
```

This will generate an `int64` columns, but all Go methods for it will accept `time.Duration` (including predicates). For example:

```go
client.FieldType.
	Query().
	Where(fieldtype.DurationLT(time.Hour)).
	Exist(ctx)
```

However, your case is a bit different, as you want to define a custom column type for your field. You can do it using the [`SchemaType`](https://entgo.io/docs/schema-fields/#database-type) option, but in order to map it to a Go type, you'll need to implement the [ValueScanner](https://pkg.go.dev/github.com/facebook/ent/schema/field#ValueScanner) interface (e.g. [`sql.NullString`](https://golang.org/pkg/database/sql/#NullString)).

**Comment by mgabeler-lee-6rs** [comment]
Yep, I managed to figure out much of this after posting this. In my case, having the field typed as `interval` at the DB layer is important, so the simpler solution you proposed won't work for me.

I was able to get things working with a custom `Interval` go type wrapping `time.Duration` however, with the `ent` schema building off `field.String`, after some hiccups:

What had me stuck initially was that the codegen doesn't handle custom `GoType`s correctly for non-struct types -- it generates `&CustomType{}` in one or two places. But a struct that embeds a `time.Duration` is easy to make work, other than actually writing the code to parse PSQL interval strings (which isn't _that_ hard, but it isn't trivial glue code either). I'll file a separate issue for that codegen problem shortly.

**Comment by wenerme** [comment]
Maybe this is about allowed the transfer form sql type to custom go type, like `sql.NullInt64` to `time.Duration`. Like `ValueScannerType` -> `GoType` mapping.

**Comment by mgabeler-lee-6rs** [comment]
Yes. The frustrating piece is that you _can't_ do `type SqlInterval time.Duration`, because `ent` codegen will try to write `foo = yourpackage.SqlInterval{}` which won't compile. You have to make it a struct, which means you need a bunch of boilerplate to wrap & unwrap that struct to get normal Duration values in and out.

**Comment by a8m** [comment]
> Yes. The frustrating piece is that you _can't_ do `type SqlInterval time.Duration`, because `ent` codegen will try to write `foo = yourpackage.SqlInterval{}` which won't compile. You have to make it a struct, which means you need a bunch of boilerplate to wrap & unwrap that struct to get normal Duration values in and out.

This can be resolved with a little change in the codegen template. I'll try to give it a look later this week.

**Comment by mgabeler-lee-6rs** [comment]
> This can be resolved with a little change in the codegen template

The changes to 0.8.0 indeed resolved this for me, and so I now have `type Interval = time.Duration` with all my custom methods attached (and `type NullInterval struct { ... }` to address the issues on #1449.

The project in which I've built these is slated to be open sourced "soon", so I'd be happy to contribute a cleaned up version of them (e.g. with reasonable MySQL support, and better SQLite support than I have now) at that point.
